pandas/core/tools/datetimes.py

```diff
@@ -140,6 +139,226 @@ def _return_parsed_timezone_results(result, timezones, box, tz):
+        format = _guess_datetime_format_for_array(arg, dayfirst=dayfirst)
+
+    if format is not None:
+        # There is a special fast-path for iso8601 formatted
+        # datetime strings, so in those cases don't use the inferred
+        # format because this path makes process slower in this
+        # special case
+        format_is_iso8601 = _format_is_iso(format)
+        if format_is_iso8601:
+            require_iso8601 = not infer_datetime_format
+            format = None
+
+    try:
+        result = None
+
+        if format is not None:
+            # shortcut formatting here
+            if format == '%Y%m%d':
+                try:
+                    result = _attempt_YYYYMMDD(arg, errors=errors)
+                except:
+                    raise ValueError("cannot convert the input to "
+                                     "'%Y%m%d' date format")
+
+            # fallback
+            if result is None:
+                try:
+                    result, timezones = array_strptime(
+                        arg, format, exact=exact, errors=errors)
+                    if '%Z' in format or '%z' in format:
+                        return _return_parsed_timezone_results(
+                            result, timezones, box, tz)
+                except tslib.OutOfBoundsDatetime:
+                    if errors == 'raise':
+                        raise
+                    result = arg
+                except ValueError:
+                    # if format was inferred, try falling back
+                    # to array_to_datetime - terminate here
+                    # for specified formats
+                    if not infer_datetime_format:
+                        if errors == 'raise':
+                            raise
+                        result = arg
+
+        if result is None and (format is None or infer_datetime_format):
+            result = tslib.array_to_datetime(
+                arg,
+                errors=errors,
+                utc=tz == 'utc',
+                dayfirst=dayfirst,
+                yearfirst=yearfirst,
+                require_iso8601=require_iso8601
+            )
+
+        if is_datetime64_dtype(result) and box:
+            result = DatetimeIndex(result, tz=tz, name=name)
+        return result
+
+    except ValueError as e:
+        try:
+            values, tz = conversion.datetime_to_datetime64(arg)
+            return DatetimeIndex._simple_new(values, name=name, tz=tz)
+        except (ValueError, TypeError):
+            raise e
+
+
+def _adjust_to_origin(arg, origin, unit):
+    """
+    Helper function for to_datetime.
+    Adjust input argument to the specified origin
+
+    Parameters
+    ----------
+    arg : list, tuple, ndarray, Series, Index
+        date to be adjusted
+    origin : 'julian' or Timestamp
+        origin offset for the arg
+    unit : string
+        passed unit from to_datetime, must be 'D'
+    Returns
+    -------
+    ndarray of adjusted dates
```

Comment:
Is it necessarily an `ndarray`?  Couldn't it be a `Timestamp`?

need newline before `Returns`

Reply:
Good catch, yeah this can be a scalar value. Will fix that tonight.